### ISSUE TITLE
Regression: `Type of "get" is partially unknown` for final TypeDicts

### ISSUE THREAD

**Issue by trygveaa** [microsoft/pyright#5452]
**Describe the bug**
With v1.1.136 I'm now getting an unknown member type when using `.get(...)` to access a member of a TypedDict. With v1.1.135 I didn't get any error.

**To Reproduce**
Run pyright on the code below. Observe this error:

```
  test.py:8:1 - error: Type of "get" is partially unknown
    Type of "get" is "Overload[(k: Literal['a']) -> str, (k: Literal['a'], default: Any) -> str, (k: LiteralString) -> None, (k: LiteralString, default: Unknown | __TDefault@Test) -> (Unknown | __TDefault@Test), (k: str) -> (Any | None), (k: str, default: Any | __TDefault@Test) -> (Any | __TDefault@Test)]" (reportUnknownMemberType)
```

**Expected behavior**
I expect no errors from this code, nothing should be unknown.

**Code or Screenshots**

```python
from typing import TypedDict, final

@final
class Test(TypedDict):
    a: str

x: Test = {"a": ""}
x.get("a")
```

**VS Code extension or command-line**
Running the command line tool v1.1.136.

**Additional context**
Note that it only happens with `@final`. If I remove that, there's no error.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by trygveaa** [comment]
Thanks!

I see that it now returns a union with `Any` if you use the default argument. Is that intentional?

E.g. if you add `reveal_type(x.get("b", 123))` to the example above. Then version 1.1.292 revealed the type `int`, while version 1.1.317 reveals the type `Any | int`.

**Comment by erictraut** [comment]
Yes, this is intentional. The change that resulted in this regression was intended to make the `get` method for `TypedDict` more consistent with the behavior of `dict`. This is a side effect of that change.

**Comment by trygveaa** [comment]
But this is now inconsistent with `dict`. If you run:

```
x = {"a": ""}
reveal_type(x.get("b", 123))
```

It reveals `str | int` as expected, without any `Any`.

**Comment by trygveaa** [comment]
I would think that if the TypedDict type is correct, we know that `b` isn't present, so the type would be `int`.

But this doesn't really affect me, only the original issue did, so not that important.

**Comment by erictraut** [comment]
This overload is required for the case where the type checker does not know the literal key value. For example:

```python
def func(x: Test, key: str):
    return x.get(key, 123)
```

This type of access is permitted for typed dictionaries, so we need to accommodate it.

**Comment by trygveaa** [comment]
Ah, thanks for the explanation. I see that if I add a `@final` decorator to the TypedDict, it also gets an overload with `LiteralString` to distinguish between those two cases though. With version 1.1.292 it then gets just the type `int` for `x.get("b", 123)` and `Any | int` for `x.get(key, 123)` as expected, but with 1.1.317 this is now `Any | int` for both.

**Comment by kylebebak** [comment]
Hey @erictraut 

Given that what I proposed in https://github.com/microsoft/pyright/issues/5883 isn't a good fit for pyright, I was hoping you could take one more look at this issue. IMO this wasn't a bug, and calling `dict.get` on a `TypedDict` with an unknown key should resolve to `Unknown` instead of `Any`.

Basically, I think the current implementation makes it hard to safely use `TypedDict`s with `NotRequired` or `total=False`, for reasons I mentioned in the other issue. I also think it doesn't fit with pyright's approach to [Unknown vs Any](https://microsoft.github.io/pyright/#/mypy-comparison?id=unknown-type-and-strict-mode).

> Pyright differentiates between explicit and implicit forms of Any. The implicit form is referred to as Unknown. For example, if a parameter is annotated as list[Any], that is a use of an explicit Any, but if a parameter is annotated as list, that is an implicit Any, so pyright refers to this type as list[Unknown]

Let's say someone writes code like this:

```py
class D(TypedDict):
    foo: str


d = D(foo="abc")
```

`a = d.get("baz")` resolves to `Any | None`, but the user hasn't done anything explicit with the keys not defined in `D`. They haven't used `Any` in the definition of `D`. I think resolution to `Any` here is implicit, and for that reason it should resolve to `Unknown` instead.

Workarounds for client code are simple. Either use `cast`, e.g. `a = cast(int | None, d.get("baz"))`, or even better, add the `baz` key to the typed dict if it might be present.

I don't think are use cases where it's safer or better for `a`'s type to resolve to `Any`. Once it resolves to `Any`, `a` can get passed anywhere and the type checker can't flag it, even with the `reportUnknown...` rules enabled.

**Comment by erictraut** [comment]
Changing the `get` method to return an `Unknown` for one of the synthesized overloads isn't an option because this would produce a false positive error when strict-mode type checking is enabled and the `get` method is accessed. This error would be reported on the member access expression `d.get` prior to the evaluation of the containing call expression `d.get("baz")` because the type of `d.get` would be "partially unknown". See above for details. So no, this isn't a viable solution to the problem you've presented.

As I explained in the other thread, the `get` method in Python is designed to be a safe accessor for a Mapping object. It is not considered an error to pass a key that does not exist, and the signature of the `get` method reflects this.

It is not the goal of a static type checker to catch every possible bug in your code. You still need to debug your code, write test cases, etc. to ensure that your code works as intended. Pyright helps catch the most common classes of bugs where it is practical to do so without generating false positive errors in the process. This particular case doesn't strike me as a class of bug that occurs frequently, nor does there appear to be a practical way to detect and report it without generating false positive errors. I'll note also that other type checkers such as mypy do not catch this case either.

**Comment by kylebebak** [comment]
Ok, if I'm understanding correctly it's just not possible to have the `dict.get` fallback overload for unknown keys resolve to `Unknown | None`, while also having the synthesized overload for known keys resolve to the known value, in a way that doesn't raise errors if the type checker is run in strict mode

Thank you again for the explanation
